### PR TITLE
Refactor to allow content_length limitation to webpreview generation

### PR DESCRIPTION
**Pull request by inkhey**
Refactoring proposal to : 
- add content_length limit that work for all requests (with and without content_length)
- have better separation between the request code and the html parsing code making easier for other program to reuse the same request to do other stuff.

This proposals:
- Don't break automated tests.
- Should not change much behavior

note: this fix also https://github.com/ludbek/webpreview/issues/15

If validated, i will propose this change upstream.

### CODE REVIEW

File: webpreview/previews.py [hunk 1]
```diff
@@ -8,38 +8,100 @@
...
+                if self.content_length_limit and length > self.content_length_limit:
+                    return None
+
+            binary_content = b''.join(chunk_data)
+        return binary_content
+
+    @property
+    def text_content(self):
+        if self.binary_content:
+            #
```
Comment (TheWathis) [inline webpreview/previews.py]:
```suggestion
```

File: webpreview/previews.py [hunk 2]
```diff
@@ -183,22 +232,28 @@ def __init__(self, *args, **kwargs):
         super(Schema, self).__init__(*args, **kwargs)
 
 
-def web_preview(url, timeout=None, headers=None, absolute_image_url=False, content=None, parser=None):
+def web_preview(url, timeout=None, headers=None, absolute_image_url=False, content=None, parser=None, content_length_limit = None):
```
Comment (TheWathis) [inline webpreview/previews.py]:
```suggestion
def web_preview(url, timeout=None, headers=None, absolute_image_url=False, content=None, parser=None, content_length_limit=None):
```

**Review by TheWathis — CHANGES_REQUESTED** [review]

File: webpreview/previews.py [hunk 3]
```diff
@@ -8,38 +8,99 @@
 from .helpers import process_image_url
 
 
+class LengthLimitedResponse(requests.Response):
```
Comment (grignards) [inline webpreview/previews.py:11]: I do not understand why this class inherits from `requests.Response` and is still given a response instance.
If it is a wrapper it shouldn't inherit I think.
Also I'm unsure a class is needed here, wouldn't be simpler to put `binary_content` + `text_content` in a simple function which is given a `request.Response` instance?

Comment (grignards) [reply webpreview/previews.py:11]: And max_content_length would be a more explicit naming I think.

File: webpreview/previews.py [hunk 4]
```diff
@@ -8,38 +8,99 @@
...
-
-        # if content is provided don't fetch from url
+    def __init__(self, url=None, properties=None, timeout=None, headers=None, content=None, parser='html.parser', content_length_limit=None):
         if not content:
-            content = PreviewBase.get_content(url, timeout, headers)
+            response = do_request(url, timeout, headers, content_length_limit)
+            content = response.text_content
+            url = response.origin_url
+        if not content:
+                raise Exception()
```
Comment (grignards) [inline webpreview/previews.py:103]: raise `Exception`? Seems strange to me.

File: webpreview/previews.py [hunk 5]
```diff
@@ -8,38 +8,99 @@
...
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+    if not valid_url.match(url):
+        raise InvalidURL("The URL is invalid.")
+
+    # if no schema add http as default
+    try:
+        if not (url.startswith('http://') or url.startswith('https://')):
+            raise MissingSchema()
+    except MissingSchema: # if no schema add http as default
+            url = "http://" + url
```
Comment (grignards) [inline webpreview/previews.py:82]: I'm unsure why is this change necessary (in regards to the original code), but removing the `try` clause would be easier to understand (raise an exception to catch it in the same block).

Comment (inkhey) [reply webpreview/previews.py:82]: Right, the original code does a request, this code avoid the need of request but you're right we can make things simpler. i just don't see it.

File: webpreview/previews.py (hunk 3 above)
Comment (inkhey) [reply webpreview/previews.py:11]: renamed to MaxLengthResponse

File: webpreview/previews.py (hunk 4 above)
Comment (inkhey) [reply webpreview/previews.py:103]: fixed with specific exception.

File: webpreview/previews.py (hunk 5 above)
Comment (inkhey) [reply webpreview/previews.py:82]: simplified.

**Pull request merged**